students/visokoo/lesson03/assignment/src/basic_operations.py

```diff
@@ -0,0 +1,161 @@
+from src.models import DATABASE, Customer, peewee
+
+logging.basicConfig(level=logging.INFO)
+LOGGER = logging.getLogger(__name__)
+
+
+def add_customer(customer_id,
+                 name,
+                 last_name,
+                 home_address,
+                 phone_number,
+                 email_address,
+                 status,
+                 credit_limit):
+    """add_customer(customer_id, name, last_name, home_address, phone_number,
+                    email_address, status, credit_limit)
+
+    Add customer info into DB based on provided params.
+
+    :param customer_id str: Unique customer ID
+    :param name str: First name of customer
+    :param last_name str: Last name of customer
+    :param str: Phone number of customer
+    :param email_address str: Email address of customer
+    :param status str: Active or Inactive customer
+    :param credit_limit int: Credit limit for customer
+
+    :return None
+    :rtype None
+    """
+    try:
+        with DATABASE.transaction():
+            new_customer = Customer.create(
+                customer_id=customer_id,
+                name=name,
+                last_name=last_name,
+                home_address=home_address,
+                phone_number=phone_number,
+                email_address=email_address,
+                status=status,
+                credit_limit=credit_limit)
+            LOGGER.info("DATABASE add successful")
+    except peewee.IntegrityError:
+        LOGGER.info("Unique constraint failed on %s", customer_id)
+
+
+def search_customer(customer_id):
+    """search_customer(customer_id)
+
+    Search for a customer via provided customer_id in DB.
+    Returns the object in dict format.
+
+    :param customer_id str: Unique customer ID
+
+    :return dict representation of existing customer, empty dict
+    if not found.
+    :rtype dict
+    """
+    query_dict = {}
+    try:
+        query = (
+            Customer
+            .select()
+            .where(Customer.customer_id == customer_id).get())
+        query_dict = model_to_dict(query)
+    except peewee.DoesNotExist:
+        LOGGER.info("Can't find customer with id: %s.", customer_id)
+        LOGGER.info("Returning empty dict.")
+    return query_dict
+
+
+def delete_customer(customer_id):
+    """delete_customer(customer_id)
+
+    Search for a customer via provided customer_id in DB.
+    Run delete_instance() on the selected item so we can cascade delete
+    if there are dependents. Will throw an error message if the customer
+    cannot be found.
+
+    Returns True if the deletion was successful.
+
+    :param customer_id str: Unique customer ID
+
+    :return Return true if the customer was deleted.
+    :rtype boolean
+    """
+    try:
+        with DATABASE.transaction():
+            query = (
+                Customer
+                .select()
+                .where(Customer.customer_id == customer_id).get())
+            query.delete_instance(recursive=True)
```

Comment:
Just for curiosity: what "recursive=True" means? Why did you use it here?

Reply:
It's an option from the delete_instance method that basically allows a cascade delete of all dependent/related rows of the one you're deleting so you're not leaving any traces.